plugins/backend/decsync/decsyncListeners.vala

```diff
@@ -0,0 +1,285 @@
+				return;
+			}
+			var feed = m_plugin.m_db.read_feed(feedID);
+			if (feed == null) return;
+			var currentCatID = feed.getCatString();
+			string newCatID;
+			if (entry.value.is_null())
+			{
+				newCatID = m_plugin.uncategorizedID();
+			}
+			else
+			{
+				newCatID = entry.value.get_string();
+			}
+			if (newCatID == null)
+			{
+				Logger.warning("Invalid catID " + Json.to_string(entry.value, false));
+				return;
+			}
+			addCategory(m_plugin, newCatID);
+			m_plugin.m_db_write.move_feed(feedID, currentCatID, newCatID);
+		}
+	}
+
+	public class CategoryNamesListener : OnSubfileEntryUpdateListener<Unit> {
+
+		private Gee.List<string> m_subfile;
+		private decsyncInterface m_plugin;
+
+		public CategoryNamesListener(decsyncInterface plugin)
+		{
+			this.m_subfile = toList({"categories", "names"});
+			this.m_plugin = plugin;
+		}
+
+		public override Gee.List<string> subfile()
+		{
+			return m_subfile;
+		}
+
+		public override void onSubfileEntryUpdate(Decsync.Entry entry, Unit extra)
+		{
+			var catID = entry.key.get_string();
+			if (catID == null)
+			{
+				Logger.warning("Invalid catID " + Json.to_string(entry.key, false));
+				return;
+			}
+			var name = entry.value.get_string();
+			if (name == null)
+			{
+				Logger.warning("Invalid name " + Json.to_string(entry.value, false));
+				return;
+			}
+			m_plugin.m_db_write.rename_category(catID, name);
+			Logger.debug("Renamed category " + catID + " to " + name);
+		}
+	}
+
+	public class CategoryParentsListener : OnSubfileEntryUpdateListener<Unit> {
+
+		private Gee.List<string> m_subfile;
+		private decsyncInterface m_plugin;
+
+		public CategoryParentsListener(decsyncInterface plugin)
+		{
+			this.m_subfile = toList({"categories", "parents"});
+			this.m_plugin = plugin;
+		}
+
+		public override Gee.List<string> subfile()
+		{
+			return m_subfile;
+		}
+
+		public override void onSubfileEntryUpdate(Decsync.Entry entry, Unit extra)
+		{
+			var catID = entry.key.get_string();
+			if (catID == null)
+			{
+				Logger.warning("Invalid catID " + Json.to_string(entry.key, false));
+				return;
+			}
+			string parentID;
+			if (entry.value.is_null())
+			{
+				parentID = CategoryID.MASTER.to_string();
+			}
+			else
+			{
+				parentID = entry.value.get_string();
+			}
+			if (parentID == null)
+			{
+				Logger.warning("Invalid parentID " + Json.to_string(entry.value, false));
+				return;
+			}
+			addCategory(m_plugin, parentID);
+			m_plugin.m_db_write.move_category(catID, parentID);
+			Logger.debug("Moved category " + catID + " to " + parentID);
+		}
+	}
+
+	private static void addCategory(decsyncInterface plugin, string catID)
+	{
+		if (catID == plugin.uncategorizedID() || catID == CategoryID.MASTER.to_string() || plugin.m_db.read_category(catID) != null)
+		{
+			return;
+		}
+		var cat = new Category(catID, catID, 0, 99, CategoryID.MASTER.to_string(), 1);
+		var list = new Gee.LinkedList<Category>();
```

Comment:
You can use `ListUtils.single(cat)` to generate a list with one element as efficiently as possible.

(Also, LinkedLists are almost always worse than ArrayLists)